### PR TITLE
Pv2 657 provider payment handle all month end payment

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs [hunk 1]
```diff
@@ -1,36 +1,89 @@
...
+                if (!isMonthEnd)
+                    return;
+
+                paymentLogger.LogVerbose($"Processing Month End for {paymentModel.GetType().Name} Ukprn {message.Ukprn} - AcademicYear {message.CollectionPeriod.AcademicYear} - Period {message.CollectionPeriod.Period}");
+
+                var monthEndJobId = await monthEndService.GetMonthEndJobId(message.Ukprn,
+                    message.CollectionPeriod.AcademicYear,
+                    message.CollectionPeriod.Period).ConfigureAwait(false);
+
+                var paymentEvent = MapToProviderPaymentEvent(message, monthEndJobId, paymentModel.EventId);
```
Comment (carlbm) [inline src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs]: Can you refactor this and the following lines with the normal `ProcessProviderMonthEndCommand` into a service

File: src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Mapping/FundingSourceToPaymentEventMappingTests.cs [hunk 2]
```diff
@@ -115,5 +113,88 @@ public void CanMapFromLevyFundingSourcePaymentEventToPayment()
...
+                IlrSubmissionDateTime = DateTime.UtcNow,
+                EventTime = DateTimeOffset.UtcNow,
+                RequiredPaymentEventId = Guid.NewGuid()
+            };
+            var payment = Mapper.Map<EmployerCoInvestedFundingSourcePaymentEvent, EmployerCoInvestedProviderPaymentEvent>(employerCoInvested);
+            payment.Ukprn.Should().Be(employerCoInvested.Ukprn);
+            payment.CollectionPeriod.Period.Should().Be(employerCoInvested.CollectionPeriod.Period);
+            payment.CollectionPeriod.AcademicYear.Should().Be(employerCoInvested.CollectionPeriod.AcademicYear);
+            payment.DeliveryPeriod.Should().Be(employerCoInvested.DeliveryPeriod);
+            //payment.FundingSourceId.Should().Be(employerCoInvested.EventId);
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Mapping/FundingSourceToPaymentEventMappingTests.cs]: Payment event doesn't have a funding source event id

[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.ProviderPayments.Application.UnitTests/Mapping/FundingSourceToPaymentEventMappingTests.cs]: line should be removed

File: src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/app.config [hunk 3]
```diff
@@ -13,7 +13,7 @@
     <add key="ProviderPaymentsServiceEndpointName" value="sfa-das-payments-providerpayments" />
     <add key="Environment" value="DEVELOPMENT" />
     <add key="TimeToWait" value="00:00:30" />
-    <add key="TimeToPause" value="00:00:01" />
+    <add key="TimeToPause" value="00:00:05" />
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/app.config]: Changes Reverted

File: src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs [hunk 4]
```diff
@@ -107,19 +138,32 @@ public async Task ThenAtMonthEndTheProviderPaymentsServiceWillPublishTheFollowin
                         && expectedEvent.Type == receivedEvent.TransactionType
                         && expectedEvent.FundingSourceType == receivedEvent.FundingSourceType
                         && expectedEvent.Amount == receivedEvent.AmountDue
-                        && TestSession.JobId == receivedEvent.JobId
+                        //&& TestSession.JobId == receivedEvent.JobId
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs]: Removed

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IMonthEndCache.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using SFA.DAS.Payments.Application.Repositories;
+using SFA.DAS.Payments.ProviderPayments.Model;
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Services
+{
+    public interface IMonthEndCache
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/IMonthEndCache.cs]: The key is currently only used by provider payment Month-end Cache so no need to expose it. by pulling the logic into a separate service.

[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/IMonthEndCache.cs]: How will you test it then?

**Comment by carlbm** [comment]
Max mentioned this on the call, but the services and handlers should be tested too

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/IlrSubmissionCache.cs [hunk 6]
```diff
@@ -19,7 +19,7 @@ public IlrSubmissionCache(IReliableStateManagerProvider stateManagerProvider, IR
             if (stateManagerProvider == null) throw new ArgumentNullException(nameof(stateManagerProvider));
             this.transactionProvider = transactionProvider ?? throw new ArgumentNullException(nameof(transactionProvider));
 
-            state = stateManagerProvider.Current.GetOrAddAsync<IReliableDictionary<string, IlrSubmittedEvent>>("SubmissionCache").Result;
+            state = stateManagerProvider.Current.GetOrAddAsync<IReliableDictionary<string, IlrSubmittedEvent>>(transactionProvider.Current, "SubmissionCache").Result;
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/IlrSubmissionCache.cs]: I will leave this, for now, can create a new task to refactor the change

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IMonthEndCache.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/IMonthEndCache.cs]: You only need to test methods that are exposed and that is covered in the Month End Service Test

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs]: Done

File: src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (carlbm) [reply src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs]: Nope...

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/ProcessProviderMonthEndCommandHandler.cs [hunk 7]
```diff
@@ -43,11 +43,13 @@ public class ProcessProviderMonthEndCommandHandler : IHandleMessages<ProcessProv
 
         public async Task Handle(ProcessProviderMonthEndCommand message, IMessageHandlerContext context)
         {
-            paymentLogger.LogDebug($"Processing Provider Month End Command for Message Id: {context.MessageId}");
+            paymentLogger.LogDebug($"Processing Provider Month End Command. Ukprn: {message.Ukprn}, Academic Year:{message.CollectionPeriod.AcademicYear}, Collection Period: {message.CollectionPeriod.Period}.");
             var currentExecutionContext = (ESFA.DC.Logging.ExecutionContext)executionContext;
             currentExecutionContext.JobId = message.JobId.ToString();
             try
             {
+                await monthEndService.StartMonthEnd(message.Ukprn, message.CollectionPeriod.AcademicYear, message.CollectionPeriod.Period, message.JobId);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/ProcessProviderMonthEndCommandHandler.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs [hunk 8]
```diff
@@ -1,36 +1,89 @@
...
         public async Task Handle(FundingSourcePaymentEvent message, IMessageHandlerContext context)
         {
-            paymentLogger.LogDebug($"Processing Funding Source Payment Event for Message Id : {context.MessageId}");
-            var paymentModel = mapper.Map<ProviderPaymentEventModel>(message);
-            await paymentsService.ProcessPayment(paymentModel, default(CancellationToken));
+            try
+            {
+                paymentLogger.LogDebug($"Processing Funding Source Payment Event for Message Id : {context.MessageId}");
+                var paymentModel = mapper.Map<ProviderPaymentEventModel>(message);
+                await paymentsService.ProcessPayment(paymentModel, default(CancellationToken));
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Handlers/FundingSourcePaymentEventHandler.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/MonthEndCache.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        public async Task<bool> Exists(long ukprn, short academicYear, byte collectionPeriod, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var key = CreateKey(ukprn, academicYear, collectionPeriod);
+            return await state.ContainsKeyAsync(transactionProvider.Current, key, TimeSpan.FromSeconds(2), cancellationToken);
+        }
+
+        public async Task<MonthEndDetails> GetMonthEndDetails(long ukprn, short academicYear, byte collectionPeriod, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var key = CreateKey(ukprn, academicYear, collectionPeriod);
+            var value = await state.TryGetValueAsync(transactionProvider.Current, key, TimeSpan.FromSeconds(2), cancellationToken);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/MonthEndCache.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/MonthEndCache.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            var key = CreateKey(ukprn, academicYear, collectionPeriod);
+            var entity = new MonthEndDetails { Ukprn = ukprn, AcademicYear = academicYear, CollectionPeriod = collectionPeriod, JobId = monthEndJobId };
+            await state.AddOrUpdateAsync(transactionProvider.Current, key, entity, (newKey, monthEnd) => monthEnd, TimeSpan.FromSeconds(4), cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<bool> Exists(long ukprn, short academicYear, byte collectionPeriod, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var key = CreateKey(ukprn, academicYear, collectionPeriod);
+            return await state.ContainsKeyAsync(transactionProvider.Current, key, TimeSpan.FromSeconds(2), cancellationToken);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/MonthEndCache.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/IlrSubmissionCache.cs [hunk 11]
```diff
@@ -39,7 +39,7 @@ public async Task AddOrReplace(string key, IlrSubmittedEvent entity, Cancellatio
 
         public async Task<ConditionalValue<IlrSubmittedEvent>> TryGet(string key, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var value = await state.TryGetValueAsync(transactionProvider.Current, key, TimeSpan.FromSeconds(4), cancellationToken);
+            var value = await state.TryGetValueAsync(transactionProvider.Current, key, TimeSpan.FromSeconds(2), cancellationToken);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService/Cache/IlrSubmissionCache.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/ProviderPaymentsService.cs [hunk 12]
```diff
@@ -38,14 +38,14 @@ public async Task ProcessPayment(ProviderPaymentEventModel payment, Cancellation
         {
             var stopwatch = new Stopwatch();
             stopwatch.Start();
-            //var isCurrentProviderIlr = await IsCurrentProviderIlr(payment.JobId, payment.Ukprn, payment.IlrSubmissionDateTime, cancellationToken);
+            var isCurrentProviderIlr = await IsCurrentProviderIlr(payment.JobId, payment.Ukprn, payment.IlrSubmissionDateTime, cancellationToken);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/ProviderPaymentsService.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 13]
```diff
@@ -1,27 +1,51 @@
...
+        }
+
+        public async Task<bool> MonthEndStarted(long ukprn, short academicYear, byte collectionPeriod)
+        {
+            return await monthEndCache.Exists( ukprn,  academicYear,  collectionPeriod);
+        }
+
+        public async Task<long> GetMonthEndJobId(long ukprn, short academicYear, byte collectionPeriod)
+        {
+            var monthEndData =  await monthEndCache.GetMonthEndDetails(ukprn, academicYear, collectionPeriod);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 14]
```diff
@@ -1,27 +1,51 @@
...
+        public async Task StartMonthEnd(long ukprn, short academicYear, byte collectionPeriod, long monthEndJobId)
+        {
+            logger.LogVerbose($"Recoding month end in the cache. Ukprn: {ukprn}, academic year: {academicYear}, collection period: {collectionPeriod}, Month End Job Id {monthEndJobId}");
+            await monthEndCache.AddOrReplace(ukprn, academicYear, collectionPeriod, monthEndJobId);
+            logger.LogDebug($"Recoded month end in the cache. Ukprn: {ukprn}, academic year: {academicYear}, collection period: {collectionPeriod} , Month End Job Id {monthEndJobId}");
+        }
+
+        public async Task<bool> MonthEndStarted(long ukprn, short academicYear, byte collectionPeriod)
+        {
+            return await monthEndCache.Exists( ukprn,  academicYear,  collectionPeriod);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 15]
```diff
@@ -1,27 +1,51 @@
...
-            CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<List<PaymentModel>> GetMonthEndPayments(CollectionPeriod collectionPeriod, long ukprn, CancellationToken cancellationToken = default(CancellationToken))
         {
             return await providerPaymentsRepository.GetMonthEndPayments(collectionPeriod, ukprn, cancellationToken);
         }
+
+        public async Task StartMonthEnd(long ukprn, short academicYear, byte collectionPeriod, long monthEndJobId)
+        {
+            logger.LogVerbose($"Recoding month end in the cache. Ukprn: {ukprn}, academic year: {academicYear}, collection period: {collectionPeriod}, Month End Job Id {monthEndJobId}");
+            await monthEndCache.AddOrReplace(ukprn, academicYear, collectionPeriod, monthEndJobId);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 16]
```diff
@@ -1,27 +1,51 @@
...
+            this.monthEndCache = monthEndCache ?? throw new ArgumentNullException(nameof(monthEndCache));
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
         }
 
-        public async Task<List<PaymentModel>> GetMonthEndPayments(CollectionPeriod collectionPeriod,
-            long ukprn,
-            CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<List<PaymentModel>> GetMonthEndPayments(CollectionPeriod collectionPeriod, long ukprn, CancellationToken cancellationToken = default(CancellationToken))
         {
             return await providerPaymentsRepository.GetMonthEndPayments(collectionPeriod, ukprn, cancellationToken);
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: done

File: src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs [hunk 17]
```diff
@@ -64,6 +67,34 @@ public async Task WhenTheFundingSourcePaymentsEventAreReceivedAsync()
             }
         }
 
+        [When(@"the period closes and month end processing begins")]
+        public async Task WhenThePeriodClosesAndMonthEndProcessingBegins()
+        {
+            await SendMonthEndEvent();
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsSteps.cs]: Done

File: src/SFA.DAS.Payments.Audit.Application.ServiceFabric/Infrastructure/StateManagerUnitOfWork.cs [hunk 18]
```diff
@@ -32,15 +32,22 @@ public Task Begin()
...
-                reliableStateManagerTransactionProvider.Current.Abort();
-                return;
+                if (ex != null)
+                {
+                    logger.LogWarning($"Rolling back the state manager transaction due to exception during message handling. Transaction Id: {reliableStateManagerTransactionProvider.Current.TransactionId}, Exception: {ex.Message}");
+                    reliableStateManagerTransactionProvider.Current.Abort();
+                    return;
+                }
+                logger.LogVerbose($"Completing state manager transaction. Transaction Id: {reliableStateManagerTransactionProvider.Current.TransactionId}");
+                await reliableStateManagerTransactionProvider.Current.CommitAsync();
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Audit.Application.ServiceFabric/Infrastructure/StateManagerUnitOfWork.cs]: Done

**Comment by samuel-dev-dba** [comment]
Closing this PR has is been merged into master via release branch

**Pull request closed without merging**